src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs

```diff
@@ -114,57 +119,44 @@ public OutputReduceToCollectionCommand CreateCommand(JsonOperationContext indexC
             return new OutputReduceToCollectionCommand(_index.DocumentDatabase, _collectionOfReduceOutputs, _reduceOutputVersion, _patternForOutputReduceToCollectionReferences, _index, indexContext, writeTxHolder);
         }
 
-        public void AddPrefixesOfDocumentsToDelete(HashSet<string> prefixes)
+        public void AddPrefixesOfDocumentsToDelete(Dictionary<string, string> prefixes)
         {
             if (_prefixesOfReduceOutputDocumentsToDelete == null)
-                _prefixesOfReduceOutputDocumentsToDelete = new ConcurrentSet<string>(StringComparer.OrdinalIgnoreCase);
+                _prefixesOfReduceOutputDocumentsToDelete = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
 
             using (_index._contextPool.AllocateOperationContext(out TransactionOperationContext context))
             using (var tx = context.OpenWriteTransaction())
             {
-                var reduceOutputsTree = tx.InnerTransaction.ReadTree(ReduceOutputsTreeName);
+                var prefixesToDeleteTree = tx.InnerTransaction.ReadTree(PrefixesOfReduceOutputDocumentsToDeleteTree);
 
                 foreach (var prefix in prefixes)
                 {
-                    if (_prefixesOfReduceOutputDocumentsToDelete.Contains(prefix))
+                    if (_prefixesOfReduceOutputDocumentsToDelete.ContainsKey(prefix.Key))
                         continue;
 
-                    reduceOutputsTree.MultiAdd(PrefixesOfReduceOutputDocumentsToDeleteKey, prefix);
+                    prefixesToDeleteTree.Add(prefix.Key, prefix.Value ?? string.Empty);
                 }
 
                 tx.Commit();
             }
 
             foreach (var prefix in prefixes)
             {
-                _prefixesOfReduceOutputDocumentsToDelete.Add(prefix);
+                _prefixesOfReduceOutputDocumentsToDelete.TryAdd(prefix.Key, prefix.Value);
             }
         }
 
-        public ConcurrentSet<string> GetPrefixesOfDocumentsToDelete()
+        public ConcurrentDictionary<string, string> GetPrefixesOfDocumentsToDelete()
         {
             return _prefixesOfReduceOutputDocumentsToDelete;
         }
 
-        public bool HasDocumentsToDelete()
-        {
-            return _prefixesOfReduceOutputDocumentsToDelete != null && _prefixesOfReduceOutputDocumentsToDelete.Count > 0;
-        }
-
         public bool HasDocumentsToDelete(TransactionOperationContext indexContext)
         {
```

Comment:
Can we rely on the value of `_prefixesOfReduceOutputDocumentsToDelete` here?
(this is now updated after the transaction was commited)

Reply:
For the sake of safety I think we're better off by relying on values which are persistent. We pass `indexContext` with open transaction so it seems to be natural to be sure that the value is as of current transaction, not the in-memory value which could be updated meanwhile (even if it's not the case in current impl).